File: java/src/test/java/io/github/messageflow/DiagnosticsTest.java
package io.github.messageflow;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.concurrent.CompletionStage;
import org.junit.jupiter.api.Test;

class DiagnosticsTest {

    private static <T> CompletionStage<T> completed(T value) {
        return CompletableFuture.completedFuture(value);
    }

    @Test
    void loggingMiddlewareRecordsStartAndCompletion() {
        var logger = new RecordingLogger(true);

        var chain = Chain.<Integer, String>create()
                .useLogging(logger, ChainLogLevel.INFORMATION, "orders")
                .useWhen(request -> true, (request, token) -> completed("handled"))
                .build();

        assertEquals("handled", chain.execute(1).toCompletableFuture().join());
        assertEquals(2, logger.entries.size());
        assertTrue(logger.entries.get(0).message.startsWith("Executing chain orders."));
        assertTrue(logger.entries.get(1).message.startsWith("Executed chain orders in "));
        assertEquals(ChainLogLevel.INFORMATION, logger.entries.get(1).level);
        assertNull(logger.entries.get(1).throwable);
    }

    @Test
    void loggingMiddlewareRecordsAsynchronousFailures() {
        var logger = new RecordingLogger(true);
        var failure = new IllegalArgumentException("boom");

        var chain = Chain.<Integer, String>create()
                .useLogging(logger)
                .useWhen(request -> true, (request, token) -> CompletableFuture.failedFuture(failure))
                .build();

        var thrown = assertThrows(
                CompletionException.class,
                () -> chain.execute(1).toCompletableFuture().join());

        assertEquals(failure, thrown.getCause());
        var last = logger.entries.get(logger.entries.size() - 1);
        assertEquals(ChainLogLevel.ERROR, last.level);
        assertEquals(failure, last.throwable);
        assertTrue(last.message.startsWith("Chain MessageFlow failed after "));
    }

    @Test
    void loggingMiddlewareRecordsSynchronousFailures() {
        var logger = new RecordingLogger(true);

        var chain = Chain.<Integer, String>create()
                .useLogging(logger)
                .use((request, next, token) -> {
                    throw new IllegalStateException("sync boom");
                })
                .build();

        var thrown = assertThrows(IllegalStateException.class, () -> chain.execute(1));

        assertEquals("sync boom", thrown.getMessage());
        var last = logger.entries.get(logger.entries.size() - 1);
        assertEquals(ChainLogLevel.ERROR, last.level);
        assertEquals(thrown, last.throwable);
    }

    @Test
    void disabledLoggerReceivesNoEntries() {
        var logger = new RecordingLogger(false);

        var chain = Chain.<Integer, String>create()
                .useLogging(logger)
                .useWhen(request -> true, (request, token) -> completed("handled"))
                .build();

        assertEquals("handled", chain.execute(1).toCompletableFuture().join());
        assertTrue(logger.entries.isEmpty());
    }

    @Test
    void loggingRejectsNullArguments() {
        var builder = Chain.<Integer, String>create();

        assertThrows(NullPointerException.class, () -> builder.useLogging(null));
        assertThrows(NullPointerException.class, () -> builder.useLogging(new RecordingLogger(true), null, "name"));
        assertThrows(
                NullPointerException.class,
                () -> builder.useLogging(new RecordingLogger(true), ChainLogLevel.DEBUG, null));
    }

    @Test
    void tracingMiddlewareCompletesTheSpan() {
        var tracer = new RecordingTracer();

        var chain = Chain.<Integer, String>create()
                .useTracing(tracer, Integer.class, String.class)
                .useWhen(request -> true, (request, token) -> completed("handled"))
                .build();

        assertEquals("handled", chain.execute(1).toCompletableFuture().join());

        var span = tracer.spans.get(0);
        assertEquals(ChainDiagnostics.EXECUTE_SPAN_NAME, span.name);
        assertEquals(Integer.class.getName(), span.requestType);
        assertEquals(String.class.getName(), span.responseType);
        assertTrue(span.ok);
        assertTrue(span.closed);
        assertNull(span.error);
    }

    @Test
    void tracingMiddlewareRecordsAsynchronousFailures() {
        var tracer = new RecordingTracer();
        var failure = new IllegalArgumentException("boom");

        var chain = Chain.<Integer, String>create()
                .useTracing(tracer, "custom", null, null)
                .useWhen(request -> true, (request, token) -> CompletableFuture.failedFuture(failure))
                .build();

        assertThrows(CompletionException.class, () -> chain.execute(1).toCompletableFuture().join());

        var span = tracer.spans.get(0);
        assertEquals("custom", span.name);
        assertEquals(failure, span.error);
        assertTrue(span.closed);
    }

    @Test
    void tracingMiddlewareRecordsSynchronousFailures() {
        var tracer = new RecordingTracer();

        var chain = Chain.<Integer, String>create()
                .useTracing(tracer, Integer.class, String.class)
                .use((request, next, token) -> {
                    throw new IllegalStateException("sync boom");
                })
                .build();

        assertThrows(IllegalStateException.class, () -> chain.execute(1));

        var span = tracer.spans.get(0);
        assertInstanceOf(IllegalStateException.class, span.error);
        assertTrue(span.closed);
    }

    @Test
    void tracingRejectsInvalidArguments() {
        var builder = Chain.<Integer, String>create();
        var tracer = new RecordingTracer();

        assertThrows(NullPointerException.class, () -> builder.useTracing(null, "name", null, null));
        assertThrows(NullPointerException.class, () -> builder.useTracing(tracer, null, null, null));
        assertThrows(IllegalArgumentException.class, () -> builder.useTracing(tracer, "", null, null));
        assertThrows(NullPointerException.class, () -> builder.useTracing(tracer, null, String.class));
        assertThrows(NullPointerException.class, () -> builder.useTracing(tracer, Integer.class, null));
    }

    @Test
    void unhandledRequestExceptionExposesItsMessageAndCause() {
        var cause = new IllegalStateException("cause");

        assertNotNull(new UnhandledRequestException().getMessage());
        assertEquals("custom", new UnhandledRequestException("custom").getMessage());
        assertEquals(cause, new UnhandledRequestException("custom", cause).getCause());
    }

    @Test
    void diagnosticConstantsAreStable() {
        assertEquals("MessageFlow", ChainDiagnostics.TRACER_NAME);
        assertEquals("MessageFlow.Execute", ChainDiagnostics.EXECUTE_SPAN_NAME);
        assertEquals("messageflow.request_type", ChainDiagnostics.REQUEST_TYPE_ATTRIBUTE);
        assertEquals("messageflow.response_type", ChainDiagnostics.RESPONSE_TYPE_ATTRIBUTE);
        assertEquals("1.0.0", ChainDiagnostics.TRACER_VERSION);
    }

    private record LogEntry(ChainLogLevel level, String message, Throwable throwable) {
    }

    private static final class RecordingLogger implements ChainLogger {

        private final boolean enabled;
        private final List<LogEntry> entries = new ArrayList<>();

        RecordingLogger(boolean enabled) {
            this.enabled = enabled;
        }

        @Override
        public boolean isEnabled(ChainLogLevel level) {
            return enabled;
        }

        @Override
        public void log(ChainLogLevel level, String message, Throwable throwable) {
            entries.add(new LogEntry(level, message, throwable));
        }
    }

    private static final class RecordingSpan implements ChainSpan {

        private final String name;
        private final String requestType;
        private final String responseType;
        private boolean ok;
        private boolean closed;
        private Throwable error;

        RecordingSpan(String name, String requestType, String responseType) {
            this.name = name;
            this.requestType = requestType;
            this.responseType = responseType;
        }

        @Override
        public void setOk() {
            ok = true;
        }

        @Override
        public void setError(Throwable throwable) {
            error = throwable;
        }

        @Override
        public void close() {
            closed = true;
        }
    }

    private static final class RecordingTracer implements ChainTracer {

        private final List<RecordingSpan> spans = new ArrayList<>();

        @Override
        public ChainSpan startSpan(String spanName, String requestType, String responseType) {
            var span = new RecordingSpan(spanName, requestType, responseType);
            spans.add(span);
            return span;
        }
    }
}

File: node/test/Diagnostics.test.ts
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import {
  createChain,
  CancellationToken,
  CancellationTokenSource,
  ChainLogLevel,
  ChainDiagnostics,
  type ChainLogger,
  type ChainTracer,
  type ChainSpan,
} from '../src/index.js';

describe('Cancellation', () => {
  it('cancellation token none is never cancelled', () => {
    const token = CancellationToken.none();
    assert.equal(token.isCancellationRequested(), false);
  });

  it('cancellation token source creates cancellable token', () => {
    const source = new CancellationTokenSource();
    const token = source.token();

    assert.equal(token.isCancellationRequested(), false);

    source.cancel();

    assert.equal(token.isCancellationRequested(), true);
  });

  it('throwIfCancellationRequested throws when cancelled', () => {
    const source = new CancellationTokenSource();
    const token = source.token();

    source.cancel();

    assert.throws(
      () => token.throwIfCancellationRequested(),
      (error: unknown) => {
        assert.ok(error instanceof Error);
        assert.equal(error.name, 'CancellationError');
        assert.match(error.message, /cancelled/i);
        return true;
      }
    );
  });

  it('throwIfCancellationRequested does not throw when not cancelled', () => {
    const token = CancellationToken.none();
    assert.doesNotThrow(() => token.throwIfCancellationRequested());
  });

  it('handlers receive cancellation token', async () => {
    const source = new CancellationTokenSource();
    let receivedToken: CancellationToken | null = null;

    const chain = createChain<number, string>()
      .useWhen(
        (_request) => true,
        (_request, token) => {
          receivedToken = token;
          return Promise.resolve('handled');
        }
      )
      .build();

    await chain.execute(1, source.token());

    assert.ok(receivedToken !== null);
    assert.equal((receivedToken as CancellationToken).isCancellationRequested(), false);

    source.cancel();
    assert.equal((receivedToken as CancellationToken).isCancellationRequested(), true);
  });

  it('execute without token uses none', async () => {
    let receivedToken: CancellationToken | null = null;

    const chain = createChain<number, string>()
      .useWhen(
        (_request) => true,
        (_request, token) => {
          receivedToken = token;
          return Promise.resolve('handled');
        }
      )
      .build();

    await chain.execute(1);

    assert.ok(receivedToken !== null);
    assert.equal((receivedToken as CancellationToken).isCancellationRequested(), false);
  });
});

describe('Logging', () => {
  class TestLogger implements ChainLogger {
    public entries: Array<{
      level: ChainLogLevel;
      message: string;
      error: unknown;
    }> = [];

    isEnabled(_level: ChainLogLevel): boolean {
      return true;
    }

    log(level: ChainLogLevel, message: string, error: unknown): void {
      this.entries.push({ level, message, error });
    }
  }

  it('logs execution start and completion', async () => {
    const logger = new TestLogger();

    const chain = createChain<number, string>()
      .useLogging(logger, ChainLogLevel.Information, 'test-chain')
      .withFallback((_request, _token) => Promise.resolve('handled'))
      .build();

    await chain.execute(1);

    assert.equal(logger.entries.length, 2);
    assert.equal(logger.entries[0].level, ChainLogLevel.Information);
    assert.match(logger.entries[0].message, /Executing chain test-chain/);
    assert.equal(logger.entries[0].error, null);

    assert.equal(logger.entries[1].level, ChainLogLevel.Information);
    assert.match(logger.entries[1].message, /Executed chain test-chain in .* ms/);
    assert.equal(logger.entries[1].error, null);
  });

  it('logs failure at error level', async () => {
    const logger = new TestLogger();
    const testError = new Error('test error');

    const chain = createChain<number, string>()
      .useLogging(logger, ChainLogLevel.Information, 'test-chain')
      .use((_request, _next, _token) => Promise.reject(testError))
      .build();

    await assert.rejects(async () => chain.execute(1));

    assert.equal(logger.entries.length, 2);
    assert.equal(logger.entries[0].level, ChainLogLevel.Information);
    assert.match(logger.entries[0].message, /Executing chain test-chain/);

    assert.equal(logger.entries[1].level, ChainLogLevel.Error);
    assert.match(logger.entries[1].message, /Chain test-chain failed after .* ms/);
    assert.equal(logger.entries[1].error, testError);
  });

  it('uses default level and name when not specified', async () => {
    const logger = new TestLogger();

    const chain = createChain<number, string>()
      .useLogging(logger)
      .withFallback((_request, _token) => Promise.resolve('handled'))
      .build();

    await chain.execute(1);

    assert.equal(logger.entries.length, 2);
    assert.equal(logger.entries[0].level, ChainLogLevel.Debug);
    assert.match(logger.entries[0].message, /MessageFlow/);
  });

  it('respects isEnabled check', async () => {
    class SelectiveLogger implements ChainLogger {
      public entries: Array<{
        level: ChainLogLevel;
        message: string;
        error: unknown;
      }> = [];

      isEnabled(level: ChainLogLevel): boolean {
        return level >= ChainLogLevel.Warning;
      }

      log(level: ChainLogLevel, message: string, error: unknown): void {
        this.entries.push({ level, message, error });
      }
    }

    const logger = new SelectiveLogger();

    const chain = createChain<number, string>()
      .useLogging(logger, ChainLogLevel.Information, 'test-chain')
      .withFallback((_request, _token) => Promise.resolve('handled'))
      .build();

    await chain.execute(1);

    assert.equal(logger.entries.length, 0);
  });

  it('throws on null logger', () => {
    const builder = createChain<number, string>();
    assert.throws(
      () => builder.useLogging(null as unknown as ChainLogger),
      TypeError
    );
  });
});

describe('Tracing', () => {
  class TestSpan implements ChainSpan {
    public status: 'ok' | 'error' | 'none' = 'none';
    public error: unknown = null;
    public closed = false;

    setOk(): void {
      this.status = 'ok';
    }

    setError(error: unknown): void {
      this.status = 'error';
      this.error = error;
    }

    close(): void {
      this.closed = true;
    }
  }

  class TestTracer implements ChainTracer {
    public spans: Array<{
      spanName: string;
      requestType: string | null;
      responseType: string | null;
      span: TestSpan;
    }> = [];

    startSpan(
      spanName: string,
      requestType: string | null,
      responseType: string | null
    ): ChainSpan {
      const span = new TestSpan();
      this.spans.push({ spanName, requestType, responseType, span });
      return span;
    }
  }

  it('creates span for successful execution', async () => {
    const tracer = new TestTracer();

    const chain = createChain<number, string>()
      .useTracing(tracer, 'test-span', 'number', 'string')
      .withFallback((_request, _token) => Promise.resolve('handled'))
      .build();

    await chain.execute(1);

    assert.equal(tracer.spans.length, 1);
    assert.equal(tracer.spans[0].spanName, 'test-span');
    assert.equal(tracer.spans[0].requestType, 'number');
    assert.equal(tracer.spans[0].responseType, 'string');
    assert.equal(tracer.spans[0].span.status, 'ok');
    assert.equal(tracer.spans[0].span.closed, true);
  });

  it('marks span as error on failure', async () => {
    const tracer = new TestTracer();
    const testError = new Error('test error');

    const chain = createChain<number, string>()
      .useTracing(tracer, 'test-span', 'number', 'string')
      .use((_request, _next, _token) => Promise.reject(testError))
      .build();

    await assert.rejects(async () => chain.execute(1));

    assert.equal(tracer.spans.length, 1);
    assert.equal(tracer.spans[0].span.status, 'error');
    assert.equal(tracer.spans[0].span.error, testError);
    assert.equal(tracer.spans[0].span.closed, true);
  });

  it('uses default span name when not specified', async () => {
    const tracer = new TestTracer();

    const chain = createChain<number, string>()
      .useTracing(tracer)
      .withFallback((_request, _token) => Promise.resolve('handled'))
      .build();

    await chain.execute(1);

    assert.equal(tracer.spans.length, 1);
    assert.equal(tracer.spans[0].spanName, ChainDiagnostics.EXECUTE_SPAN_NAME);
    assert.equal(tracer.spans[0].requestType, null);
    assert.equal(tracer.spans[0].responseType, null);
  });

  it('throws on null tracer', () => {
    const builder = createChain<number, string>();
    assert.throws(
      () => builder.useTracing(null as unknown as ChainTracer),
      TypeError
    );
  });

  it('throws on empty span name', () => {
    class EmptyTracer implements ChainTracer {
      startSpan(): ChainSpan {
        return new TestSpan();
      }
    }

    const builder = createChain<number, string>();
    assert.throws(() => builder.useTracing(new EmptyTracer(), ''), TypeError);
  });

  it('throws when tracer returns null', async () => {
    class NullTracer implements ChainTracer {
      startSpan(): ChainSpan {
        return null as unknown as ChainSpan;
      }
    }

    const chain = createChain<number, string>()
      .useTracing(new NullTracer())
      .withFallback((_request, _token) => Promise.resolve('handled'))
      .build();

    await assert.rejects(async () => chain.execute(1), TypeError);
  });
});

describe('ChainDiagnostics', () => {
  it('exposes diagnostic constants', () => {
    assert.equal(ChainDiagnostics.TRACER_NAME, 'MessageFlow');
    assert.equal(ChainDiagnostics.TRACER_VERSION, '1.0.0');
    assert.equal(ChainDiagnostics.EXECUTE_SPAN_NAME, 'MessageFlow.Execute');
    assert.equal(
      ChainDiagnostics.REQUEST_TYPE_ATTRIBUTE,
      'messageflow.request_type'
    );
    assert.equal(
      ChainDiagnostics.RESPONSE_TYPE_ATTRIBUTE,
      'messageflow.response_type'
    );
  });
});
